packages/design-system/src/DatePicker/DatePicker.tsx

```diff
@@ -0,0 +1,319 @@
+import React, { useState } from "react";
+import BaseDatePicker from "react-datepicker";
+import range from "lodash/range";
+import getYear from "date-fns/getYear";
+import getMonth from "date-fns/getMonth";
+import clsx from "classnames";
+
+import "react-datepicker/dist/react-datepicker.css";
+import "./styles.css";
+import {
+  DatePickerCalenderClassName,
+  DatePickerCalenderHeaderClassName,
+  DatePickerClassName,
+  DateTimePickerClassName,
+} from "./DatePicker.constants";
+import {
+  DatePickerHeaderProps,
+  DatePickerProps,
+  DateRangePickerProps,
+} from "./DatePicker.types";
+import { StyledDatePickerHeader } from "./DatePicker.styles";
+import { Input } from "Input";
+import { Button } from "Button";
+import { Menu, MenuContent, MenuItem, MenuTrigger } from "Menu";
+
+function DatePicker(props: DatePickerProps) {
+  const {
+    calendarClassName,
+    className,
+    inputProps,
+    inputSize = "md",
+    isClearable,
+    isDisabled,
+    isReadOnly,
+    isRequired,
+    label,
+    onChange,
+    placeholderText = "Select date",
+    selected,
+    yearEndRange,
+    yearStartRange,
+    ...rest
+  } = props;
+  const [selectedDate, setSelectedDate] = useState<Date | null>(
+    selected || null,
+  );
+
+  const onChangeHandler = (
+    date: Date | null,
+    e: React.SyntheticEvent<any, Event> | undefined,
+  ) => {
+    setSelectedDate(date);
+    onChange && onChange(date, e);
+  };
+
+  return (
+    <BaseDatePicker
+      {...rest}
+      calendarClassName={clsx(DatePickerCalenderClassName, calendarClassName, {
+        [DateTimePickerClassName]: rest.showTimeInput,
+      })}
+      className={clsx(className, DatePickerClassName)}
+      customInput={
+        <Input
+          {...inputProps}
+          endIcon={
+            isClearable && selectedDate ? "close-circle-line" : undefined
+          }
+          endIconProps={{
```

Comment:
Didn't we decide we would never pass an onClick to an icon? If it needs an onClick then it should be an iconButton no?

Reply:
we cannot use icon button here as the same as I mentioned for serach input. The variations and hover states and all needs to be sorted.

Reply:
Let's add a TODO here. https://theappsmith.slack.com/archives/D03K37HRU00/p1678879256621369?thread_ts=1678872894.044499&cid=D03K37HRU00